### PR TITLE
Adds missing schemas + data for metrics to snowplow stats ping

### LINKED ISSUE #48424: Add Metrics to snowplow stat ping
# Add Metrics info to Snowplow Stats Ping

The details are [in Notion](https://www.notion.so/metabase/Migrate-Anonymous-Stats-ping-to-Snowplow-18f155e235d34ba09f45518a866336e2?pvs=4#77eb4c912bd449b78b6b72a88d61a329).

This work will be adding a the metrics section to the stats ping.

### PR DESCRIPTION
**Pull request by escherize**
resolves #48424

Most of the info for this section was in the legacy stats info. The new stuff is calculated in [`->snowplow-metric-info`](https://github.com/metabase/metabase/blob/add-metrics-to-snowplow-ping/src/metabase/analytics/stats.clj#L599).

The specified tags are included with `"metrics"`.

Also, I refactored this to be more repl-friendly. We can call `generate-instance-stats!` from the repl to see its value.

This adds the jsonschema for `"metrics"` section, as well as `"grouped_metrics"`.

### CODE REVIEW

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:63]: **nitpick (blocking):** We need to make this a new minor schema version `1-0-1`.

Context: In SnowcatCloud the schemas cannot be edited when they are deployed already. This means if we change it here and start sending new properties (i.e. `metrics`), the event will be rejected. So if we do updates after deployment, they need to be new schema versions that we then deploy again in SnowcatCloud 🙂

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:52]: **question:** What would be the cases when the value is `null`? Shouldn't we always get a value for a metric?

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:44]: **question:** 

Since I cannot find it yet: How did we decide about the `tags` property which was listed [here](https://www.notion.so/metabase/Migrate-Anonymous-Stats-ping-to-Snowplow-18f155e235d34ba09f45518a866336e2?pvs=4#e3d7fdc32d3446df9e486ce8d63da04b)?

**Review by Somtom — COMMENTED** [review]
Really excited to see our progress on this migration. I left a few questions and one blocking nitpick that should be easy to resolve :)

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:63]: Thank you for pointing this out! 

I will revert this file, and add the `"metrics"` key to an identical json file in `metabase/snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1`. 

I think I also need to bump the value for `::instance_stats` to `"1-0-1"` [here](https://github.com/metabase/metabase/blob/add-metrics-to-snowplow-ping/src/metabase/analytics/snowplow.clj#L35-L41). But I'm not sure: Do you know when a new version of a schema gets added to SnowcatCloud? Are there some docs for this? I'll look around when I get a chance.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:52]: For the `"metrics"` section, I can be sure we don't produce nulls, so I'll remove it.

Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:52]: Fixed

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:44]: Thanks for the pointer. I don't see any tags in the product doc, but I will add support for them to the schema and stats collecting code.

Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0]: Done.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (luizarakaki) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:38]: Should we add grouped_metrics too?

Comment (luizarakaki) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:38]: So we don't need to change the schema again to add group_metrics

Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:38]: good call, I'll add it to the schema

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (Somtom) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0]: > I think I also need to bump the value for ::instance_stats to "1-0-1"
Yes I think that, too (but @qnkhuat) might have better knowledge and can confirm.
[Here](https://www.notion.so/metabase/Snowplow-integration-5da1f874beda4153b4fccfa6c1e77caa?pvs=4#723b7505593c4f7fa1e29416a2377ffa) is a bit of docs around the process but I think we need to still improve it a bit (feel free to adjust it based on your learnings). 

I adjusted a small [section](https://www.notion.so/metabase/Snowplow-integration-5da1f874beda4153b4fccfa6c1e77caa?pvs=4#50e90bb17b27409da20a68a48251bf42) to give some indication on when to talk to SnowcatCloud. TL;DR: When you have the PR up and we internally agreed on the schema format, we should already sanity check with them before merging. Sometimes Joao has some last minute concerns. I was pushing for us (data team) having more ownership and being able to update the schemas. That worked for a week and now we don't have the permissions anymore since their system is not very error proof sometimes.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (Somtom) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0]: Thanks for adding!
Yeah I think there have been many docs in the meanwhile. FYI [This](https://www.notion.so/metabase/Migrate-Anonymous-Stats-ping-to-Snowplow-18f155e235d34ba09f45518a866336e2?pvs=4#e3d7fdc32d3446df9e486ce8d63da04b) was the one I was referring to.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:63]:
```suggestion
          "key": {
            "description": "The unique name of the metric",
            "type": "string",
            "maxLength": 255
          },
          "value": {
            "description": "The value of the metric",
            "type": ["number"],
            "maxLength": 255,
            "minimum": 0,
            "maximum": 2147483647
          },,
          "tags": {
            "description": "Tags that can be used flagging teams / features the metric belongs to",
            "type": "array",
            "items": {
              "type": "string",
              "maxLength": 255
            }
          }
        },
        "required": ["name", "value"]
```

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:159]: Should we change this as well to make the new fields required?

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (qnkhuat) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:8]: version here is mismtached, should be `1-0-1` right?

Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1]: fixed.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:63]: This almost all looks fine, except that we need to remember that `cache_num_queries_cached` is actually a string value because it is a binned balue. We need to either change how it's computed or allow strings in here. (example strings are "0", "1-5", etc. from [here](https://github.com/metabase/metabase/blob/add-metrics-to-snowplow-ping/src/metabase/analytics/stats.clj#L54).

So, we could just return the number and not bin it.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:159]: Well, we are adding the _schema_ for grouped metrics here, but not the data for it. so I will make "metrics" required. but we should hold off on "grouped_metrics" (which will be addable w/o updating the schema in a followup).

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:63]: > `"required": ["name", "value"]`

The maps in this array have the keys: "key" "value" and "tags". I don't think we can require a "name" key w/o specifying what it is.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:35]: Looking into this some more, we probably want to say that "grouped_metrics" is a required key. So we will have to change this some more. I'll leave in the "grouped_metrics" schema, however once we are sending `grouped_metrics`, I recommend that we alter [snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1](https://github.com/metabase/metabase/pull/48476/files/4e6afe06fd061a5af249fc7745a76e6ef004c456#diff-830f3618a350a9306a891e1713625bab0c5e8d030fdbced4be15544ee9ed4044) line 159 to be 

```diff 
- "required": ["instance_attributes", "features", "metadata", "metrics"]
+ "required": ["instance_attributes", "features", "metadata", "metrics", "grouped_metrics"]
```

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:63]: I've made this into `"required": ["key", "value", "tags"]` since they will always have those keys.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0]: fixed. Thanks for the links

**Comment by escherize** [comment]
I decided that it's better to add the `"grouped_metrics"` schema when we start sending grouped metrics.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:35]: Yes. Note that if you make this change after the schema has been deployed, then this means we need to create another schema version. 

We should actually double-check with Joao from SnowcatCloud whether adding new required fields works with a minor version upgrade or whether this would be a major one. If major, we should maybe already add it as required fields and send empty data or something.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1]: >So, we could just return the number and not bin it.

Yes agreed. If we had binned stuff and care about the counts per bin, this would more become a grouped metric. In this case it sounds reasonable to just always return numbers for metrics.

> The maps in this array have the keys: "key" "value" and "tags". I don't think we can require a "name" key w/o specifying what it is.

Ah yeah sorry - good catch. I forgot to change the `key` to `name`. I think in the schema from the docs, we agreed on `name` for metric instead of `key`.  Similar to how we do it for grouped metrics. There you also have a `name`.
Does that make more sense with that context?

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1]: Yep, saw it in the [other discussion](https://github.com/metabase/metabase/pull/48476#discussion_r1795761878) and commented there. 
The approach is fine for me as long as we don't run into issues such that we need to create a major version update for our schema (that always requires more work since it creates a new table in data land). I should have thought about this earlier when we created the first initial version of the schema. Let's double-check with Joao from SnowcatCloud.

**Review by Somtom — COMMENTED** [review]
Thanks man for tackling the comments. I tried to add a bit of clarification for some things. I really love the progress I see. 
There is one small unknown regarding the required fields that we should figure out and a few minor nits but I am already very happy :)

**Review by noahmoss — APPROVED** [review]

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1]: I believe we should not include grouped metrics until they're being created. We will still have to update the schema (to make them required, and change the type from `[object, null]`, to `object`. I have a copy of the grouped_metrics json schema handy and I can paste it into an issue for adding grouped_metrics.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1]: Cool, I will switch it to "name" "value" and "tags", and unbin `cache_num_wueries_cached`.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:35]: see: https://github.com/metabase/metabase/pull/48476#issuecomment-2405802467

I decided that it's better to add the "grouped_metrics" schema when we start sending grouped metrics. We would still have to update the schema to make the grouped metrics section required once implemented. This way we can setup the schema, and make it required at the same time.

**Comment by escherize** [comment]
I've tested this with snowplow mini, and it accepts the value 🎉.

Open question:

- According to the snowplow docs, adding a required field (as we are in this PR) is a breaking change and should mean bumping to 2-0-0. However I also read that we generally shouldn't need to do that, so I could use some clarity there.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:8]: After [discussion](https://metaboat.slack.com/archives/C04FZTRF457/p1728889142471029) with Joao, it seems a better and safer approach to go with `2-0-0`. Let's change this here and also rename the file.

Apart from that the file looks good to me. I also asked Joao to double-check whether it is valid (in the linked thread above)

File: src/metabase/analytics/snowplow.clj
Comment (Somtom) [inline src/metabase/analytics/snowplow.clj:47]: When we rename to `2-0-0` we also should not forget to adjust here

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:210]: Based on the discussion with Joao (making this 2-0-0), we should be safe keeping them required. We should just make sure to send empty arrays for the attributes where we don't have the logic implemented yet.

**Review by Somtom — COMMENTED** [review]
Sorry for the back and forth but there were new infos based on the discussion with SnowcatCloud

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (Somtom) [inline snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:16]: https://metaboat.slack.com/archives/C04FZTRF457/p1728979421380599?thread_ts=1728889142.471029&cid=C04FZTRF457

File: src/metabase/analytics/stats.clj
Comment (albertoperdomo) [inline src/metabase/analytics/stats.clj:609]: We've since broken this setting up into multiple ones. This seems outdated. No?

**Review by albertoperdomo — CHANGES_REQUESTED** [review]

**Comment by albertoperdomo** [comment]
Should this not be backported to the 51 release branch so it makes it into 51? I see the no-backport label.

File: src/metabase/analytics/stats.clj
Comment (Somtom) [reply src/metabase/analytics/stats.clj:609]: I think this is according to the [docs](https://www.notion.so/metabase/Migrate-Anonymous-Stats-ping-to-Snowplow-18f155e235d34ba09f45518a866336e2?pvs=4#1414ef173a184569a2278760650971eb) or what specifically are you referring to @albertoperdomo ?

Comment (albertoperdomo) [reply src/metabase/analytics/stats.clj:609]: There is no setting enable-embedding anymore, we broke this up into three (static, interactive and SDK). See https://github.com/metabase/metabase/pull/48154

Comment (albertoperdomo) [reply src/metabase/analytics/stats.clj:609]: I will update the doc.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:16]: done

File: src/metabase/analytics/stats.clj
Comment (escherize) [reply src/metabase/analytics/stats.clj:609]: @albertoperdomo I see how it looks similar to the setting, but that is not the setting, it is the column on the dashboard table. It hasn't changed.

**Comment by escherize** [comment]
> Should this not be backported to the 51 release branch so it makes it into 51? I see the no-backport label.

Thanks, I've switched the labels so we will backport it.

File: src/metabase/analytics/snowplow.clj
Comment (escherize) [reply src/metabase/analytics/snowplow.clj:47]: done

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1]: sounds good, done.

File: snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1
Comment (escherize) [reply snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1]: done.

**Review by albertoperdomo — APPROVED** [review]

**Pull request merged**
